src/lib/tls/tls13/tls_handshake_layer_13.cpp

```diff
@@ -0,0 +1,197 @@
+   {
+   m_read_buffer.insert(m_read_buffer.end(), data_from_peer.cbegin(), data_from_peer.cend());
+   }
+
+namespace {
+
+constexpr size_t HEADER_LENGTH = 4;
+
+template<typename Msg_Type>
+Handshake_Type handshake_type_from_byte(uint8_t type)
+   {
+   if constexpr(std::is_same_v<Msg_Type, Handshake_Message_13>)
+      {
+      switch(type)
+         {
+         case CLIENT_HELLO:
+         case SERVER_HELLO:
+         case END_OF_EARLY_DATA:
+         case ENCRYPTED_EXTENSIONS:
+         case CERTIFICATE:
+         case CERTIFICATE_REQUEST:
+         case CERTIFICATE_VERIFY:
+         case FINISHED:
+            return Handshake_Type(type);
+         }
+      throw TLS_Exception(Alert::UNEXPECTED_MESSAGE, "Unknown handshake message received");
+      }
+   else
+      {
+      switch(type)
+         {
+         case NEW_SESSION_TICKET:
+         case KEY_UPDATE:
+            return Handshake_Type(type);
+         }
+      throw TLS_Exception(Alert::UNEXPECTED_MESSAGE, "Unknown post-handshake message received");
+      }
+   }
+
+template<typename Msg_Type>
+std::optional<Msg_Type> parse_message(TLS::TLS_Data_Reader& reader, const Policy& policy,
+                                      const Connection_Side peer_side)
+   {
+   // read the message header
+   if(reader.remaining_bytes() < HEADER_LENGTH)
+      { return std::nullopt; }
+
+   Handshake_Type type = handshake_type_from_byte<Msg_Type>(reader.get_byte());
+
+   // make sure we have received the full message
+   const size_t msg_len = reader.get_uint24_t();
+   if(reader.remaining_bytes() < msg_len)
+      { return std::nullopt; }
+
+   // create the message
+   const auto msg = reader.get_fixed<uint8_t>(msg_len);
+   if constexpr(std::is_same_v<Msg_Type, Handshake_Message_13>)
+      {
+      switch(type)
+         {
+         case CLIENT_HELLO:
+            return Client_Hello_13(msg);
+         case SERVER_HELLO:
+            // SERVER_HELLO might be either an actual server_hello (1.2 or 1.3) or a
+            // hello_retry_request. Hence, this construction is exceptionally
+            // funneled through a factory method and then transformed into a
+            // generic Handshake_Message_13.
+            return std::visit([](auto message) -> Handshake_Message_13
+               { return message; }, Server_Hello_13::parse(msg));
+         // case END_OF_EARLY_DATA:
+         //    return End_Of_Early_Data(msg);
+         case ENCRYPTED_EXTENSIONS:
+            return Encrypted_Extensions(msg);
+         case CERTIFICATE:
+            return Certificate_13(msg, policy, peer_side);
+         // case CERTIFICATE_REQUEST:
+         //    return Certificate_Req_13(msg);
+         case CERTIFICATE_VERIFY:
+            return Certificate_Verify_13(msg, peer_side);
+         case FINISHED:
+            return Finished_13(msg);
+         default:
+            BOTAN_ASSERT(false, "cannot be reached"); // make sure to update handshake_type_from_byte
```

Comment:
If the messages commented above are received, this _will_ be reached. Handle NYI messages better.

Reply:
Fixed in https://github.com/randombit/botan/pull/2922/commits/8976bf250a3e7c299a4b36485f0328a843f2627e by commenting the culprit message types in `handshake_type_from_byte()` (invoked above) resulting in an UNEXPECTED_MESSAGE alert instead of failed assertion here.

The unimplemented messages are currently not negotiated in the Client Hello, so a compliant server will not send them. Except the server might still request client authentication (sending a Certificate Request message) which we would ungracefully reject right now. Client authentication will come in the next PR today or tomorrow, though.